### PR TITLE
ext_proc: allow override metadata without grpc_service

### PR DESCRIPTION
**Pull request by sitano**
Commit Message: `ext_proc: allow override metadata without grpc_service`

The patch introduces `overrides.grpc_initial_metadata` in `ExtProcPerRoute.overrides` that appends and overrides parent metadata:

      - match:
          ...
        route:
          ...
        typed_per_filter_config:
          envoy.filters.http.ext_proc:
            "@type": type.googleapis.com/envoy.extensions.filters.http.ext_proc.v3.ExtProcPerRoute
            overrides:
              grpc_initial_metadata:
                - key: "b"
                  value: "c"
                - key: "c"
                  value: "c"
    http_filters:
    - name: envoy.filters.http.ext_proc
    typed_config:
      "@type": type.googleapis.com/envoy.extensions.filters.http.ext_proc.v3.ExternalProcessor
      grpc_service:
        envoy_grpc:
          cluster_name: ext-proc-proxy
        timeout: 60s
        initial_metadata:
          - key: "a"
            value: "a"
          - key: "b"
            value: "b"

that allows to have:

    a: a
    b: c
    c: c

like the following:

    [2024-01-09 16:40:23.241][1967519][debug][router] [source/common/router/router.cc:731] [Tags: "ConnectionId":"0","StreamId":"5894068517272876876"] router decoding headers:
    ':method', 'POST'
    ':path', '/envoy.service.ext_proc.v3.ExternalProcessor/Process'
    ':authority', 'ext-proc-proxy'
    ':scheme', 'http'
    'te', 'trailers'
    'content-type', 'application/grpc'
    'a', 'a'
    'b', 'c'
    'c', 'c'
    'x-envoy-internal', 'true'
    'x-forwarded-for', '172...'

Additional Description: `new ext_proc per route override of inherited grpc metadata without specifying grpc_service`
Risk Level: low
Testing: Unit, Integration, Config
Docs Changes: will have to update https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto#extensions-filters-http-ext-proc-v3-extprocperroute regarding `repeated config.core.v3.HeaderValue grpc_initial_metadata = 7;`. 

> // Additional metadata to include into streams initiated to the ext_proc gRPC
> // service. This can be used for scenarios in which additional ad hoc        
> // authorization headers (e.g. ``x-foo-bar: baz-key``) are to be injected or 
> // when a route needs to partially override inherited metadata.              

Release Notes: see `changlelogs/current.yaml`
Platform Specific Features: -

Testing
===

    $ bazel test //test/extensions/filters/http/ext_proc:config_test --config=clang --verbose_failures --test_arg="-l trace"
    $ bazel test //test/extensions/filters/http/ext_proc:filter_test --config=clang --verbose_failures --test_arg="-l trace"
    $ bazel test //test/extensions/filters/http/ext_proc:ext --config=clang --verbose_failures --test_arg=--gtest_filter='*/ExtProcIntegrationTest.PerRouteGrpcMetadata/*' --test_arg="-l trace"

Test infrastructure https://github.com/sitano/envoy/tree/ivan_api_502/examples/grpc-bridge:

    t1$ cd ./server && go run
    t2$ cd ./ext_proc && go run
    t3$ envoy/bazel-bin/source/exe/envoy-static --config-path ./envoy-proxy.yaml --log-level debug --base-id 1
    // t4$ envoy/bazel-bin/source/exe/envoy-static --config-path ./envoy-proxy.yaml --log-level debug --base-id 2
    t5$ envoy/bazel-bin/source/exe/envoy-static --config-path ./envoy-proxy-ext_proc-2.yaml --log-level trace --base-id 3
    $ CLIENT_PROXY=http://127.0.0.1:9911 ./client.py set key1 1
    $ CLIENT_PROXY=http://127.0.0.1:9911 ./client.py set key2 1
    $ CLIENT_PROXY=http://127.0.0.1:9911 ./client.py set key3 1
    $ ss -tnp

Example
===

- [ext_proc_metadata_per_route_example.yaml](https://github.com/envoyproxy/envoy/files/13885400/ext_proc_metadata_per_route_example.txt)

### CODE REVIEW

**Comment by sitano** [comment]
Other considerations:

1. probably `FilterConfigPerRoute::merge` code maybe simplified to something like replace_all or push_back, but I am not sure what minimal version of C++ we are supporting.
2. I am not sure about `context.messageValidationVisitor()` in the config test.
3. Speaking of integration test, I did not find in a couple of hours a way to catch the ext_proc processor stream metadata during `sendDownstreamRequest()`. Maybe someone would be kind to give suggestions if there is sense to extend it further.

Another PR working on ext_proc metadata, but unrelated: https://github.com/envoyproxy/envoy/pull/30747

**Comment by sitano** [comment]
Also, I have full grpc exp_proc example in https://github.com/sitano/envoy/tree/ivan_api_502/examples/grpc-bridge which I would like to add too, but it will a bit more effort to polish and I think its worth putting it into a separate PR. Or we can add it here?

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:280]: It is unclear to me which field this is overriding?
Also, are these essentially request-headers that need to be added?

**Review by adisuissa — COMMENTED** [review]
Thanks!
Left an API high-level comment.
Please note that this PR will be reviewed after marking it as non-draft (otherwise the reviewers may not be notified on its changes).

**Comment by sitano** [comment]
This test `Mobile/Compile time options` is also finished but did not update the status (https://github.com/envoyproxy/envoy/actions/runs/7463682673).

It is ready for review. The PR touches the same areas as the aforementioned https://github.com/envoyproxy/envoy/pull/30747 and it could be easier for the same reviewers to do the review as they already are familiar with the ext_proc code base.

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (sitano) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:278]: for more info check the example https://github.com/envoyproxy/envoy/files/13885400/ext_proc_metadata_per_route_example.txt

Comment (adisuissa) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:278]: Thanks for explaining this.
The `ExtProcOverrides` type allows overriding fields that are in `ExternalProcessor` for a specific route. For example, if the route needs a different processing_mode, the a value in [ExtProcOverrides->processing_mode](https://github.com/envoyproxy/envoy/blob/143ef4a82d26a75ccd460563f5e48044607883c5/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L255) will override the value from the [ExternalProcesseor->processing_mode](https://github.com/envoyproxy/envoy/blob/143ef4a82d26a75ccd460563f5e48044607883c5/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L119).

IIUC the `initial_metadata` field you are referring to is inside the [GrpcService type](https://github.com/envoyproxy/envoy/blob/234dcc646cbfc1cd512742fc92ad704f514b5948/api/envoy/config/core/v3/grpc_service.proto#L302). If that's the case, why not override the [grpc_service](https://github.com/envoyproxy/envoy/blob/143ef4a82d26a75ccd460563f5e48044607883c5/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L272) field?

Comment (sitano) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:278]: > IIUC the initial_metadata field you are referring to is inside the [GrpcService type](https://github.com/envoyproxy/envoy/blob/234dcc646cbfc1cd512742fc92ad704f514b5948/api/envoy/config/core/v3/grpc_service.proto#L302). If that's the case, why not override the [grpc_service](https://github.com/envoyproxy/envoy/blob/143ef4a82d26a75ccd460563f5e48044607883c5/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L272) field?

@adisuissa it is because when you do provide `ExtProcPerRoute.overrides.grpc_service` [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto#extensions-filters-http-ext-proc-v3-extprocperroute)] you 

1. can't provide only  `ExtProcPerRoute.overrides.grpc_service.initial_metadata` - it is necessary to provide at least one of `grpc_service.envoy_grpc` or `grpc_service.google_grpc`.
2. even if we could do (1) the original semantics of `ExtProcPerRoute.overrides.grpc_service` is that it is completely REPLACES the original inherited configuration from the ExternalProcessor [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto#extensions-filters-http-ext-proc-v3-externalprocessor)] whilst we need to only extend and maybe override - basically inherit the parents configuration with the small changes.
3. even if (2) was possible to do with just `ExtProcPerRoute.overrides.grpc_service` without the issue of (1) the `grpc_service.initial_metadata` name suggests that it does not have the extending semantics but the replaсement. 
4. then GrpcService [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/grpc_service.proto#config-core-v3-grpcservice)] is the global configuration component reused in many places of configuration so we can't change its semantics. as well as no one would appreciate the change in the semantics of `ExtProcPerRoute.overrides.grpc_service.initial_metadata` if we had tried to modify it.

We needed the option to extend (A) and modify (B) inherited (C) metadata configuration for _ext_proc_ service on the per route basis WITHOUT overriding the GrpcService [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/grpc_service.proto#config-core-v3-grpcservice)] configuration itself - that is why we extended the `ExtProcPerRoute.overrides` with the `ExtProcPerRoute.overrides.metadata` repeatable field. It was impossible to do before this patch.

With this patch we now can have a big parent ExternalProcessor [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto#extensions-filters-http-ext-proc-v3-externalprocessor)] configuration with proper GrpcService [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/grpc_service.proto#config-core-v3-grpcservice)] and corresponding `initial_metadata` and only extend it in our thousands of routes.

Comment (adisuissa) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:278]: The `ExtProcOverride` semantics is always replacement.
From an API perspective, I'm concerned about the case where both `grpc_service` and `metadata` are overridden. In this case, it is not clear what the `initial_metadata` should be.

Is (1) you stated above the main issue you see in using the current `grpc_service` override option?

I think another option could be adding a new type that customizes the GrpcService, where other features can be added (not only overriding the `initial_metadata`, but other fields as needed). @envoyproxy/api-shepherds may have a different perspective on this.

Comment (sitano) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:278]: > The ExtProcOverride semantics is always replacement.

In this patch, we allowed for whatever `initial_metadata` configuration there is to converge first (merge parent ExternalProcessor with ExtProcPerRoute that actually replaces the `grpc_service` configuration completely with the latest variant), and only then be extended with the `ExtProcOverride.metadata`.

> From an API perspective, I'm concerned about the case where both grpc_service and metadata are overridden. In this case, it is not clear what the initial_metadata should be.

This has been resolved surprisingly well. See the example [ext_proc_metadata_per_route_example.yaml](https://github.com/envoyproxy/envoy/files/13885400/ext_proc_metadata_per_route_example.txt) the part with `key3` (`CLIENT_PROXY=http://127.0.0.1:9911 ./client.py set key3 1`):

When you have this:

```
              - match:
                  prefix: "/kv.KV"
                  headers:
                  - name: "key"
                    exact_match: "key3"
                route:
                  cluster: backend-proxy
                typed_per_filter_config:
                  envoy.filters.http.ext_proc:
                    "@type": type.googleapis.com/envoy.extensions.filters.http.ext_proc.v3.ExtProcPerRoute
                    overrides:
                      grpc_service:
                        envoy_grpc:
                          cluster_name: ext-proc-proxy
                        timeout: 30s
                        initial_metadata:
                          - key: "c"
                            value: "c"
                          - key: "d"
                            value: "d"
                      metadata:
                        - key: "d"
                          value: "blah"
                        - key: "g"
                          value: "blah"
          http_filters:
          - name: envoy.filters.http.ext_proc
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.ext_proc.v3.ExternalProcessor
              failure_mode_allow: false
              async_mode: false
              allow_mode_override: true
              message_timeout: 360s
              grpc_service:
                envoy_grpc:
                  cluster_name: ext-proc-proxy
                timeout: 60s
                initial_metadata:
                  - key: "a"
                    value: "a"
                  - key: "b"
                    value: "b"
```

you will get

```
    [debug][router] [source/common/router/router.cc:731] [Tags: "ConnectionId":"0","StreamId":"5894068517272876876"] router decoding headers:
    ':method', 'POST'
    ':path', '/envoy.service.ext_proc.v3.ExternalProcessor/Process'
    ':authority', 'ext-proc-proxy'
    ':scheme', 'http'
    'te', 'trailers'
    'content-type', 'application/grpc'
    'c', 'c'
    'd', 'blah'
    'g', 'blah'
    'x-envoy-internal', 'true'
    'x-forwarded-for', '172...'
```

The `ExtProcPerRoute.overrides.grpc_service.initial_metadata` replaces the original `ExternalProcessor.grpc_service.initial_metadata` and then is perfectly extended by the `ExtProcOverride.metadata`.

> Is (1) you stated above the main issue you see in using the current grpc_service override option?

Mostly because of (2). `ExtProcOverride.grpc_service` completely replaces parent `grpc_service`.

**Comment by ravenblackx** [comment]
Added reviewer and approver from #30747 as suggested.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (sitano) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: @yanavlasov I have moved the things.

Also looked around and have got 2 considerations regarding code in ext_proc.cc:

- the neighbor `init*` and `merge*` (ie `initGrpcService`) functions are class members. I don't see much sense in adding `initMetadata` and `mergeHeaderValue` to the class private definition as well as that those aforementioned ones are there. but having mine in an anonymous namespace adds a little inconsistency to the local source code.  shall we move our functions to the class members as well? or maybe move all inits and merges out of the class members?
- ext_proc.cc has some static functions (like `immediateResponseChecker()`, `allDisabledMode()`), do we want them to be moved to an anonymous namespace as well? maybe not in this PR...

**Comment by yanavlasov** [comment]
LGTM from me. I will wait for API LGTM.

/wait-any

**Review by tyxia — COMMENTED** [review]
@sitano  Just one high-level question, Can I understand as follow?
Before this PR, ext_proc overrides whole grpc_service
with this PR, you want to only overrides `initial_metadata` of the grpc_service, rather than whole gRPC_service

**Comment by sitano** [comment]
> Before this PR, ext_proc overrides whole grpc_service

@tyxia yes - replaces completely with the instance from ExtProcPerRoute.

> with this PR, you want to only overrides initial_metadata of the grpc_service, rather than whole gRPC_service

yes - override inherited metadata - the metadata data of the converged grpc_service configuration.

**Comment by yanavlasov** [comment]
@sitano please also merge main to resolve merge conflict

**Comment by sitano** [comment]
@yanavlasov merged

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (tyxia) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:278]: I am wondering if we should name it `grpc_initial_metadata` explicitly? So that it is clear that this field is used for partial grpc service override, although it is clearly described in comment.
I don't feel strongly and will let @adisuissa @yanavlasov  and other folks make the call.

**Comment by yanavlasov** [comment]
LGTM from me. I will wait for API LGTM and @tyxia final pass.

/wait-any

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:908]: style nit: 
Instead of having multiple anonymous namespaces in this file, can we move those functions into one? And please move anonymous namespace to the top below  `namespace ExternalProcessing {`  or after using declaration if you need any.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:156]: nit: `bool found = false` .

You only need a boolean flag here to set and check

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:133]: No need to have this function. I think you can do the initialization in constructor init list like one line below: 
```
metadata_(config.overrides().metadata().begin(), config.overrides().metadata().end()
```

**Review by tyxia — COMMENTED** [review]
LGTM, modulo some nits

Also, as mentioned in this [thread](https://github.com/envoyproxy/envoy/pull/31544/files#r1459288352), I personally prefer substitute `metadata` with `grpc_initial_metadata` in this PR as it is more clear since there are many kinds of metadata in envoy. I don't feel strongly if other reviewers think it is ok.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (sitano) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:908]: @tyxia would you prefer me to move funcs only that I have added or all the static ones that are in that file?

Comment (sitano) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:908]: I can also make some inits and merge functions static as they do not use *this.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (sitano) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: done

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (sitano) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: I have moved everything up into one anon namespace and along with moving out some functions out of the class members.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (sitano) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: done

**Review by tyxia — APPROVED** [review]
LGTM, Thanks!

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (sitano) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: fixed

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (sitano) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: 👍  I replaced those with `constexpr absl::string_view` but had to add one ABSL_ATTRIBUTE_UNUSED for FilterName because it is not used anywhere. With std::string without constexpr it had generated init code that was referencing the global variable but with constexpr absl::string_view it's now clearly unreferenced and need unused. I don't think we also want to get rid of this var.

File: changelogs/current.yaml
Comment (phlax) [inline changelogs/current.yaml:26]:
```suggestion
    config API to allow extending inherited metadata from ``ExternalProcessor``
```

Comment (sitano) [reply changelogs/current.yaml]: changed

Comment (phlax) [reply changelogs/current.yaml]: @sitano - huge apology here - after i commented i realized its a name rather than a literal - ideally link to the api doc i think

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:41]: If it is not used,  i think we can just remove it in this PR

File: changelogs/current.yaml
Comment (sitano) [reply changelogs/current.yaml]: @phlax - no worries. changed to 
```
...
config API to allow extending inherited metadata from
:ref:`ExternalProcessor.grpc_service <envoy_v3_api_field_extensions.filters.http.ext_proc.v3.ExternalProcessor.grpc_service>`
and 
:ref:`ExtProcOverrides.grpc_service <envoy_v3_api_field_extensions.filters.http.ext_proc.v3.ExtProcOverrides.grpc_service>`
with the new or updated values.
```
to emphasize that the metadata can be inherited from either option.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (sitano) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: removed

**Comment by tyxia** [comment]
@sitano  For the build error, looks like you need to sync with latest main, `FilterName` is used in https://github.com/envoyproxy/envoy/pull/31090 which was just merged two days ago

**Comment by sitano** [comment]
@tyxia thanks for pointing 31090. Updated main.

**Review by tyxia — APPROVED** [review]
LGTM, Thanks

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
@adisuissa please take a look at API

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (adisuissa) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:275]: The issue with merge-semantics is that it hinders the understandability of the config.
The filter-config override can appear in multiple-levels: Virtual-Host, Route, and Cluster-Weight, which make the merge semantics unclear.
cc @envoyproxy/api-shepherds for additional views.

Comment (adisuissa) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:275]: > I am wondering if we should name it `grpc_initial_metadata` explicitly? So that it is clear that this field is used for partial grpc service override, although it is already described in comment. I don't feel strongly and will let @adisuissa @yanavlasov and other folks make the call.

If the API field remains then yeah the field should be exact as possible so one can understand exactly which metadata field is set by it.

**Comment by yanavlasov** [comment]
@sitano  please resolve merge conflicts

File: changelogs/current.yaml
Comment (yanavlasov) [inline changelogs/current.yaml:37]:
```suggestion
    :ref:`metadata <envoy_v3_api_field_extensions.filters.http.ext_proc.v3.ExtProcOverrides.grpc_metadata>`
```

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: changelogs/current.yaml
Comment (sitano) [reply changelogs/current.yaml]: sorry, missed that one. fixed.

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (sitano) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto]: as we have https://github.com/envoyproxy/envoy/pull/30747 merged first with `metadata_options = 6;` I have renamed this PR field into `repeated config.core.v3.HeaderValue grpc_metadata = 7;`. Now it must be much more clear what metadata it is capable of extending.

**Comment by sitano** [comment]
list of additional changes  in this `main` update:

- `override.metadata` became `override.grpc_metadata` - must be much less confusing about what metadata it is extending
- `initNamespaces, initUntypedForwardingNamespaces, initTypedForwardingNamespaces,initUntypedReceivingNamespaces` made static into anonymous namespace
-  `initNamespaces` has removed unnecessary (IMHO) early exit path: 

```
-  if (ns.empty()) {
-    return {};
-  }
```

**Comment by htuch** [comment]
I think this also has the issue I point to in https://github.com/envoyproxy/envoy/issues/32125#issuecomment-1920526718. I'm not sure if we want to have multiple different ways to do the same thing - when to use metadata vs. attributes? 

They both sound like words for OOB properties, is it possible to formally state when you would use one vs the other? 

CC @rshriram

**Comment by sitano** [comment]
> I'm not sure if we want to have multiple different ways to do the same thing - when to use metadata vs. attributes?

@htuch attributes of the https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto (request_attributes) (see also https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/attributes#arch-overview-request-attributes) are completely different from the gRPC service metadata. 

> is it possible to formally state when you would use one vs the other?

in this PR we are only providing a way to extend grpc service initial_metadata on per route basis.

**Comment by sitano** [comment]
EDIT of last msg:

> If I understand the point in 32125 the question is that existing attributes are not flexible enough and serve the same goal as metadata. I also think it is better to only have 1 thing. However, I would expect we can not deprecate the metadata from the grpc_service - it is used in many other places. Then why extend attributes functionality? It looks like it makes sense but it is out of the scope of this PR.

**Comment by htuch** [comment]
Let's continue this discussion in https://github.com/envoyproxy/envoy/issues/32125. I don't think there is a problem with making initial metadata per-route, if we go a certain direction #32125, then this is somewhat redundant and I think we would do many of the things we do with metadata via attributes. Let's discuss.

**Comment by alyssawilk** [comment]
@htuch I'm not clear on the status of tihs PR.  is it awaiting review or waiting on offline discussion?

**Comment by sitano** [comment]
As initial metadata stays as per https://github.com/envoyproxy/envoy/issues/32125#issuecomment-1930556312 I think we can proceed?

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:310]: This PR is fine from an API perspective, will defer to others to merge. One slight nit though, I probably would make this `grpc_initial_metadata` if it was me given all the other metadata options and contexts floating around.

Comment (sitano) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto]: 👍  - renamed to `grpc_initial_metadata`

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (sitano) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: changed to `has_key/s`

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (sitano) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto]: we now have it as `grpc_initial_metadata`

**Comment by kyessenov** [comment]
I'm also confused about the primary approver for this PR. @yanavlasov @htuch @rshriram let's clarify the status, is this mostly ready and awaiting for nits, or major changes needed.

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by htuch** [comment]
API LGTM, handing back to @yanavlasov for implementation review and merge.

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
Comment (jbohanon) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:2609]: I don't think we're actually validating anything with this test. These checks make sure that the header value is set
```suggestion
  processRequestHeadersMessage(*grpc_upstreams_[0], true, absl::nullopt);
  EXPECT_EQ("c", processor_stream_->headers().get(Http::LowerCaseString("b"))[0]->value().getStringView());
  EXPECT_EQ("c", processor_stream_->headers().get(Http::LowerCaseString("c"))[0]->value().getStringView());

  handleUpstreamRequest();
```

Comment (sitano) [reply test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc]: nice! done!

**Review by jbohanon — APPROVED** [review]
Awesome work!

**Comment by sitano** [comment]
@yanavlasov I am not sure if we would like to have it merged, but just pinging you because > yanavlasov requested changes, and there are no changes actually to do

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
